### DISCUSSION TITLE
Service user access_token is not a JWT

### DISCUSSION THREAD

**Discussion by corneleberle** [zitadel/zitadel#1780] [Q&A] [answered]
Hi ZITADEL team

I tried to follow your guide to create an access_token for a service user (https://docs.zitadel.ch/docs/guides/usage/serviceusers).

The resulting access_token is not a JWT and therefore I can't use it for authentication on a third party API. It works perfectly fine to call one of the ZITADEL APIs (e.g. /userinfo). But it fails when I try to use the access_token as a Bearer token and call our own API.

Is it possible to get a JWT access_token for service users?

**Comment by livio-a** [comment, marked as answer]
Hi @corneleberle 

At the moment we only support opaque tokens for service users.

As defined by OAuth2 specification, a `Bearer` token is an opaque string and APIs should treat it that way. Even when using JWT as access_token, APIs can only really make sure the access_token is valid by using the OAuth introspection endpoint. Validating the signature, checking expiration, issued_at and so on will not guarantee that. The token might have been revoked explicitly, the user might have signed out and so on.

But wee are aware that a lot of APIs are using JWT as tokens and are doing "normal" JWT validation as just described.
We will add an option in the future to get JWT for service user as well.

Is the API you're trying to call, implemented by your team or is it a real third-party API?

**Reply by corneleberle** [reply]
Hi @livio-a 

It's about an API which is implemented by our team.

I agree that if you only check if a JWT access token has a correct signature and is not expired, you don't know if it has been revoked. That's the tradeoff. Otherwise you have to introspect the token for every single request which has a negative impact on the performance.

I was trying to use the introspection endpoint to check the validity of a service user token. No luck so far. I always get a `{ "active": false }` response. If I introspect a regular user access token it works for both authentication modes (`client_secret_basic` and `private_key_jwt`).

**Reply by fforootd** [reply]
> I agree that if you only check if a JWT access token has a correct signature and is not expired, you don't know if it has been revoked. That's the tradeoff. Otherwise you have to introspect the token for every single request which has a negative impact on the performance.

Just as a side note, you can cache the response in your project for a defined amount of time, to mitigate this.
This also enables you to verify different calls with different "security" guarantees.

**Reply by livio-a** [reply]
The introspection endpoint checks the audience of the access_token and if the api is part of it (either its client_id or the id of the parent project).

When authenticating to an application we set the client_id and the projectID as audience of the access_token. So direct access from an app to the api in the same project is granted without further configuration.

A service user on the other hand is not part of any project. Therefore you will have to request the projectID (of your api) to be added as audience. In ZITADEL you can do this by providing the scope `urn:zitadel:iam:org:project:id:{projectid}:aud` when authenticating the service user.
See https://docs.zitadel.ch/docs/apis/openidoauth/scopes#reserved-scopes for more information.

I see that we still have some catching up to do in this area of ​​the documentation.

**Reply by corneleberle** [reply]
Thanks for your hint. The additional `urn:zitadel:iam:org:project:id:{projectid}:aud` scope was indeed still missing.

Maybe a note in the service user guide (https://docs.zitadel.ch/docs/guides/usage/serviceusers#3-with-this-jwt-request-an-oauth-token-from-zitadel) would help to prevent others from doing the same mistake.
